File: 2022/Qround/chain.py
import sys
import heapq

def solution(data, pos):
    lookup = {}
    for i, v in enumerate(pos):
        lookup[v-1] = lookup.get(v-1, [])
        lookup[v-1].append(i)
    
    heaplookup = {}
    stack = [-1]
    while len(stack) > 0:
        last = stack.pop()
        if lookup.get(last):
            if heaplookup.get(lookup[last][0]):
                heaplookup[last] = list(heapq.merge(
                    *[ heaplookup[x] for x in lookup[last] ]
                ))
                if last == -1:
                    continue
                if data[last] > heaplookup[last][0]:
                    heapq.heappushpop(heaplookup[last], data[last])
            else:
                stack.append(last)
                for v in lookup[last]:
                    stack.append(v)
        else:
            heaplookup[last] = []
            heapq.heappush(heaplookup[last], data[last])

    return sum(heaplookup.get(-1, []))

testcase = sys.stdin.readline()
for i in range(int(testcase)):
    sys.stdin.readline()
    line1 = sys.stdin.readline().rstrip('\n')
    line2 = sys.stdin.readline().rstrip('\n')
    ans = solution(
        [ int(x) for x in line1.split(' ') ],
        [ int(x) for x in line2.split(' ') ],
    )
    print("Case #{}: {}".format(i+1, ans))

File: 2022/Qround/chain.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func toIntArr(s string) []int {
	temp := strings.Split(strings.TrimSpace(s), " ")
	out := make([]int, len(temp))
	for i, v := range temp {
		value, _ := strconv.Atoi(v)
		out[i] = value
	}
	return out
}

func solution(data, pos []int) int {
	lookup := map[int][]int{}
	for i, v := range pos {
		if lookup[v-1] == nil {
			lookup[v-1] = []int{}
		}
		lookup[v-1] = append(lookup[v-1], i)
	}
	return 0
}

func main() {
	var buf string
	stdin := bufio.NewReader(os.Stdin)
	buf, _ = stdin.ReadString('\n')
	testcase, _ := strconv.Atoi(strings.TrimSpace(buf))
	for i := 0; i < testcase; i++ {
		stdin.ReadString('\n')
		line1, _ := stdin.ReadString('\n')
		line2, _ := stdin.ReadString('\n')
		result := solution(
			toIntArr(line1),
			toIntArr(line2),
		)
		fmt.Printf("Case: %d: %d\n", i+1, result)
	}
}
